// @sapphire/plugin-api/dist/lib/structures/Route.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route = void 0;
const pieces_1 = require("@sapphire/pieces");
const discord_js_1 = require("discord.js");
const RouteData_1 = require("../utils/RouteData");
const HttpMethods_1 = require("./http/HttpMethods");
/**
 * @since 1.0.0
 */
class Route extends pieces_1.Piece {
    constructor(context, options = {}) {
        super(context, options);
        /**
         * (RFC 7230 3.3.2) The maximum decimal number of octets.
         */
        Object.defineProperty(this, "maximumBodyLength", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The accepted content types.
         */
        Object.defineProperty(this, "acceptedContentMimeTypes", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The route information.
         */
        Object.defineProperty(this, "router", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The methods this route accepts.
         */
        Object.defineProperty(this, "methods", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new discord_js_1.Collection()
        });
        const api = this.container.server.options;
        this.router = new RouteData_1.RouteData(`${api.prefix ?? ''}${options.route ?? this.name ?? ''}`);
        for (const [method, symbol] of HttpMethods_1.methodEntries) {
            const value = Reflect.get(this, symbol);
            if (typeof value === 'function')
                this.methods.set(method, value);
        }
        this.maximumBodyLength = options.maximumBodyLength ?? api.maximumBodyLength ?? 1024 * 1024 * 50;
        this.acceptedContentMimeTypes = options.acceptedContentMimeTypes ?? api.acceptedContentMimeTypes ?? null;
    }
    /**
     * Per-piece listener that is called when the piece is loaded into the store.
     * Useful to set-up asynchronous initialization tasks.
     */
    onLoad() {
        const store = this.store;
        for (const [method, cb] of this.methods) {
            store.table.get(method).set(this, cb.bind(this));
        }
        return undefined;
    }
    /**
     * Per-piece listener that is called when the piece is unloaded from the store.
     * Useful to set-up clean-up tasks.
     */
    onUnload() {
        const store = this.store;
        for (const [method] of this.methods) {
            store.table.get(method).delete(this);
        }
        return undefined;
    }
}
exports.Route = Route;
